datafusion/src/execution/memory_manager.rs

```diff
@@ -270,25 +270,35 @@ impl MemoryManager {
                     requesters: Arc::new(Mutex::new(HashSet::new())),
                     pool_size,
                     requesters_total: Arc::new(Mutex::new(0)),
-                    trackers_total: Arc::new(Mutex::new(0)),
+                    trackers_total: AtomicUsize::new(0),
                     cv: Condvar::new(),
                 })
             }
         }
     }
 
     fn get_tracker_total(&self) -> usize {
-        *self.trackers_total.lock().unwrap()
+        self.trackers_total.load(Ordering::SeqCst)
     }
 
     pub(crate) fn grow_tracker_usage(&self, delta: usize) {
-        *self.trackers_total.lock().unwrap() += delta;
+        self.trackers_total.fetch_add(delta, Ordering::SeqCst);
     }
 
     pub(crate) fn shrink_tracker_usage(&self, delta: usize) {
-        let mut total = self.trackers_total.lock().unwrap();
-        assert!(*total >= delta);
-        *total -= delta;
+        let update =
+            self.trackers_total
+                .fetch_update(Ordering::SeqCst, Ordering::SeqCst, |x| {
```

Comment:
Would this potentially end up in an infinite loop if `x` is zero as the closure would always return None?

After seeing this way, I think I agree that the original implementation using `Mutex` was better -- sorry about that @yjshen  I didn't realize the subtleties involved here

Reply:
I think fetch_update won't let infinite loop here. A simple case zero minus ten results in underflow:
https://play.rust-lang.org/?version=stable&mode=release&edition=2021

I agree the mutex version is easier to write and reason. I can revert the last commit if preferred.

Reply:
This PR is ready to go I think -- we can always keep iterating in the next one. Sorry for the delay in merging @yjshen